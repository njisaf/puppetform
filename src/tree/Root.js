import {
    types,
    getSnapshot,
} from 'mobx-state-tree';

import ErrorCollection from './ErrorCollection';
import MapUtils from './utils/MapUtils';

import LaneStore from './LaneStore';
import CardStore from './CardStore';

const Root = types.model('Root', {
        loading: false,
        laneStore: types.optional(types.compose('laneStore', LaneStore, ErrorCollection, MapUtils), {}),
        cardStore: types.optional(types.compose('cardStore', CardStore, ErrorCollection, MapUtils), {}),
    })
    .actions(self => {

        return {
            afterCreate() {
                // console.log('Initializing Root', getSnapshot(self))
            },

            initializeBoard(cards, lanes) {
                lanes.forEach((lane) => {
                    try {
                        self.laneStore.set(lane.id, lane);
                    } catch (e) {
                        self.laneStore.failure(e, lane);
                    }
                })

                cards.forEach((card) => {
                    try {
                        self.cardStore.set(card.id, card);
                    } catch (e) {
                        self.cardStore.failure(e, card);
                    }
                })
            },

            setLoading(boolean) {
                self.loading = boolean;
            },
        }
    })
    .views(self => {
        return {
            get isLoading() {
                return self.loading === true;
            },
        }
    })

export default Root;