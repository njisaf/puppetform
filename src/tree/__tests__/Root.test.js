import { getSnapshot, onSnapshot } from 'mobx-state-tree';
import Root from '../Root';

import Cards from '../__mocks__/cards.mock.json';
import Lanes from '../__mocks__/lanes.mock.json';

describe('Root', () => {
    it('can toggle .loading to true and false', () => {
        const item = Root.create();
        const states = [];
        onSnapshot(item, snapshot => {
            states.push(snapshot)
        })
        
        expect(item.loading).toBe(false);

        item.setLoading(true);
        expect(item.loading).toBe(true);
        expect(item.isLoading).toBe(true);

        expect(getSnapshot(item)).toMatchSnapshot();
        expect(states).toMatchSnapshot();
    })

    it('receives an array of lanes with attached cards, and adds both cards and lanes to the board', () => {
        const item = Root.create();
        const states = [];
        onSnapshot(item, snapshot => {
            states.push(snapshot)
        })

        expect(item.laneStore.size).toBe(0);
        item.initializeBoard(Cards, Lanes);

        expect(item.laneStore.size).toBeGreaterThan(0);

        expect(states).toMatchSnapshot();
    })
})