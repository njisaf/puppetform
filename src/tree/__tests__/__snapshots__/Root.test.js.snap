// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Root can toggle .loading to true and false 1`] = `
Object {
  "cardStore": Object {
    "failures": Object {},
    "map": Object {},
  },
  "laneStore": Object {
    "failures": Object {},
    "map": Object {},
  },
  "loading": true,
}
`;

exports[`Root can toggle .loading to true and false 2`] = `
Array [
  Object {
    "cardStore": Object {
      "failures": Object {},
      "map": Object {},
    },
    "laneStore": Object {
      "failures": Object {},
      "map": Object {},
    },
    "loading": true,
  },
]
`;

exports[`Root receives an array of lanes with attached cards, and adds both cards and lanes to the board 1`] = `
Array [
  Object {
    "cardStore": Object {
      "failures": Object {},
      "map": Object {
        "Card1": Object {
          "description": "Can AI make memes",
          "id": "Card1",
          "label": "30 mins",
          "title": "Write Blog",
        },
        "Card2": Object {
          "description": "Transfer via NEFT",
          "id": "Card2",
          "label": "5 mins",
          "title": "Pay Rent",
        },
        "Card3": Object {
          "description": "Nacho Friday",
          "id": "Card3",
          "label": "15 mins",
          "title": "Purchase Fifty Pounds of Grade A Pork Butt",
        },
      },
    },
    "laneStore": Object {
      "failures": Object {},
      "map": Object {
        "lane1": Object {
          "cards": Array [
            "card1",
            "card2",
          ],
          "id": "lane1",
          "label": "2/2",
          "title": "Planned Tasks",
        },
        "lane2": Object {
          "cards": Array [
            "card2",
            "card3",
          ],
          "id": "lane2",
          "label": "0/0",
          "title": "Completed",
        },
      },
    },
    "loading": false,
  },
]
`;
